### PR TITLE
Introducing Recommender handlers - collaborative filtering and popularity based

### PR DESCRIPTION
**Pull request by dusvyat**
## Description

handler for lightfm Recommender package - wip:
1. [x] support for user-item recommendations
2. [x] support for item-item recommendations

- [x] handler for polars based popularity recommender

# todo 

- [x] more documentation - both inline and readme format
- [x] handler for collaborative filtering
- [x] tests for collaborative filtering

## Type of change

- [x] ⚡ New feature (non-breaking change which adds functionality)
- [ ] 📄 This change requires a documentation update

### What is the solution?

Here is the proposed high level architecture:

Further documentation can be found [here](https://www.notion.so/usqrd/MindsDB-Recommender-System-package-9071d98bb1d64abf8b8372cf1027efb1?pvs=4)

## Checklist:

- [x] My code follows the style guidelines(PEP 8) of MindsDB.
- [x] I have commented my code, particularly in hard-to-understand areas.
- [ ] I have updated the documentation, or created issues to update them.
- [x] I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [ ] I have shared a short loom video or screenshots demonstrating any new functionality.

### CODE REVIEW

**Review by paxcema — CHANGES_REQUESTED** [review]
There are a few blockers before we can merge the PR, but this is awesome @dusvyat, great work! 

Also, apologies for the delayed review :sweat_smile:

File: mindsdb/integrations/handlers/lightfm_handler/lightfm_handler.py
Comment (dusvyat) [reply mindsdb/integrations/handlers/lightfm_handler/lightfm_handler.py:55]: Ok I will add a module to dataprep_ml for recommender data preprocessing perhaps

File: tests/unit/executor_test_base.py
Comment (dusvyat) [reply tests/unit/executor_test_base.py:420]: yes sorry about that, I only added the handlers to the mock db

File: mindsdb/integrations/handlers/lightfm_handler/lightfm_handler.py
Comment (dusvyat) [reply mindsdb/integrations/handlers/lightfm_handler/lightfm_handler.py:55]: I left it in for now, will remove once https://github.com/mindsdb/dataprep_ml/pull/24 is merged and a new v is released

**Comment by ZoranPandovski** [comment]
recheckcla

**Comment by mindsdbadmin** [comment]
**CLA Assistant Lite bot:** <br/>Thank you for your submission, we really appreciate it. Like many open-source projects, we ask that you all sign our [Contributor License Agreement](https://github.com/mindsdb/mindsdb_native/blob/stable/assets/contributions-agreement/individual-contributor.md) before we can accept your contribution. You can sign the CLA by just posting a Pull Request Comment same as the below format.<br/>
   - - -
   I have read the CLA Document and I hereby sign the CLA
   - - -
   **13** out of **15** committers have signed the CLA.<br/>:white_check_mark: @KenmogneThimotee<br/>:white_check_mark: @yusharth<br/>:white_check_mark: @panos-span<br/>:white_check_mark: @aia<br/>:white_check_mark: @ShruAgarwal<br/>:white_check_mark: @martyna-mindsdb<br/>:white_check_mark: @gantanikhilraj<br/>:white_check_mark: @1ce-one<br/>:white_check_mark: @marsidmali<br/>:white_check_mark: @ArmanChand<br/>:white_check_mark: @mit-27<br/>:white_check_mark: @sid2002CN<br/>:white_check_mark: @2170chm<br/>:x: @tmichaeldb<br/>:x: @dusvyat<br/><sub>You can retrigger this bot by commenting **recheck** in this Pull Request</sub>

File: tests/unit/ml_handlers/test_lightfm.py
Comment (dusvyat) [reply tests/unit/ml_handlers/test_lightfm.py:145]: perhaps we could look to pass that in via where at predict stage as opposed to when model is created? That way we can ensure we don't unnecessarily retrain models as you rightly pointed out.

Comment (dusvyat) [reply tests/unit/ml_handlers/test_lightfm.py:146]: Ok I have done that, very good catch. It is way better now with this change :)

As for the test I would rather leave as is as I prefer to keep tests self contained where ever possible. They also execute very fast as data used is small

File: mindsdb/integrations/handlers/lightfm_handler/lightfm_handler.py
Comment (dusvyat) [inline mindsdb/integrations/handlers/lightfm_handler/lightfm_handler.py:102]: With this change, the user will now pass in recommender_type at prediction rather than at create. 

This is much more efficient as model will only need to be fitted once for a given data source and user can then query based on type of predictions they are after.

File: tests/unit/ml_handlers/test_lightfm.py
Comment (dusvyat) [reply tests/unit/ml_handlers/test_lightfm.py:146]: @paxcema let me know if any issues with this?

File: mindsdb/integrations/handlers/lightfm_handler/lightfm_handler.py
Comment (dusvyat) [inline mindsdb/integrations/handlers/lightfm_handler/lightfm_handler.py:86]: @paxcema using evaluation model from lightfm, let me know if any issues with this

**Comment by dusvyat** [comment]
recheck

File: tests/unit/ml_handlers/test_lightfm.py
Comment (paxcema) [reply tests/unit/ml_handlers/test_lightfm.py:109]: Yep those are good points, let's leave them as separate tests 👍

File: mindsdb/integrations/handlers/lightfm_handler/lightfm_handler.py
Comment (paxcema) [reply mindsdb/integrations/handlers/lightfm_handler/lightfm_handler.py:80]: Sounds great!

**Review by paxcema — APPROVED** [review]

**Pull request merged**
